internal/graph/check.go

```diff
@@ -0,0 +1,500 @@
+// it is used to resolve union.
+func any(ctx context.Context, requests []ReduceableCheckFunc) (*dispatchpb.DispatchCheckResponse, error) {
+
+	if len(requests) == 0 {
+		// return empty CheckResult
+	}
+
+	responseMetadata := emptyMetadata
+	resultChan := make(chan *dispatchpb.DispatchCheckResponse, len(requests))
+	errChan := make(chan error)
+	defer close(resultChan)
+	defer close(errChan)
+
+	cctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	g := new(errgroup.Group)
+
+	for _, req := range requests {
+		fn := func() error {
+			err := req(cctx, resultChan)
+			if err != nil {
+				errChan <- err
+			}
+
+			return err
+		}
+
+		g.Go(fn)
+	}
+
+	for i := 0; i < len(requests); i++ {
+		select {
+		case result := <-resultChan:
+			responseMetadata = combineResponseMetadata(responseMetadata, result.Metadata)
+
+			if result.WrappedResponse.Allowed {
+				return &dispatchpb.DispatchCheckResponse{
+					Metadata: responseMetadata,
+					WrappedResponse: &openfgapb.CheckResponse{
+						Allowed: true,
+					},
+				}, nil
+			}
+		case err := <-errChan:
+			return &dispatchpb.DispatchCheckResponse{}, err
+		case <-ctx.Done():
+			// return err
+		}
+	}
+
+	return &dispatchpb.DispatchCheckResponse{
+		Metadata: responseMetadata,
+		WrappedResponse: &openfgapb.CheckResponse{
+			Allowed: false,
+		},
+	}, nil
+}
+
+// difference defines a CheckReducer that returns a check result indicating if the base check passes and none of
+// the subsequent checks pass (the subtracted set).
+func difference(ctx context.Context, requests []ReduceableCheckFunc) (*dispatchpb.DispatchCheckResponse, error) {
+	cctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if len(requests) < 2 {
+		// return a non-nil error, because you need a base and subtract reducer
+	}
+
+	responseMetadata := emptyMetadata
+	baseChan := make(chan *dispatchpb.DispatchCheckResponse, 1)
+	subtractChan := make(chan *dispatchpb.DispatchCheckResponse, len(requests)-1)
+	errChan := make(chan error, 1)
+	defer close(baseChan)
+	defer close(subtractChan)
+	defer close(errChan)
+
+	g := new(errgroup.Group)
+
+	g.Go(func() error {
+		err := requests[0](cctx, baseChan)
+		if err != nil {
+			errChan <- err
+		}
+
+		return err
+	})
+
+	for _, req := range requests[1:] {
+		fn := func() error {
+			err := req(cctx, subtractChan)
+			if err != nil {
+				errChan <- err
+			}
+
+			return err
+		}
+
+		g.Go(fn)
+	}
+
+	for i := 0; i < len(requests); i++ {
+		select {
+		case base := <-baseChan:
+			responseMetadata = combineResponseMetadata(responseMetadata, base.Metadata)
+
+			if !base.WrappedResponse.Allowed {
+				return &dispatchpb.DispatchCheckResponse{
+					Metadata: responseMetadata,
+					WrappedResponse: &openfgapb.CheckResponse{
+						Allowed: false,
+					},
+				}, nil
+			}
+		case sub := <-subtractChan:
+			responseMetadata = combineResponseMetadata(responseMetadata, sub.Metadata)
+
+			if sub.WrappedResponse.Allowed {
+				return &dispatchpb.DispatchCheckResponse{
```

Comment:
I plan on cleaning up these responses. Instead of inline instantiating the response objects, I plan on invoking a helper that will make reading this easier.

For example, `checkResponse(metadata, allowed) *dispatchpb.DispatchCheckResponse`